### ISSUE TITLE
Gerrit permission issues

### ISSUE THREAD

**Issue by HenryWinterbottom-NOAA** [NOAA-EMC/GSI#165]
There currently seems to be a Gerrit permissions issue for some of us within ESRL/PSD. The following is the error:

```
Cloning into '/scratch2/BMC/gsienkf/Henry.Winterbottom/regression-tests/builds/UFS-RNR.20210524/sorc/ufsrnr-gsi/fix'...
ssh: connect to host vlab.ncep.noaa.gov port 29418: Connection refused
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: clone of 'gerrit:GSI-fix' into submodule path '/scratch2/BMC/gsienkf/Henry.Winterbottom/regression-tests/builds/UFS-RNR.20210524/sorc/ufsrnr-gsi/fix' fai
led
Failed to clone 'fix'. Retry scheduled
Cloning into '/scratch2/BMC/gsienkf/Henry.Winterbottom/regression-tests/builds/UFS-RNR.20210524/sorc/ufsrnr-gsi/libsrc'...
ssh: connect to host vlab.ncep.noaa.gov port 29418: Connection refused
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: clone of 'gerrit:GSI-libsrc' into submodule path '/scratch2/BMC/gsienkf/Henry.Winterbottom/regression-tests/builds/UFS-RNR.20210524/sorc/ufsrnr-gsi/libsr
c' failed
Failed to clone 'libsrc'. Retry scheduled
Cloning into '/scratch2/BMC/gsienkf/Henry.Winterbottom/regression-tests/builds/UFS-RNR.20210524/sorc/ufsrnr-gsi/fix'...
ssh: connect to host vlab.ncep.noaa.gov port 29418: Connection refused
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
fatal: clone of 'gerrit:GSI-fix' into submodule path '/scratch2/BMC/gsienkf/Henry.Winterbottom/regression-tests/builds/UFS-RNR.20210524/sorc/ufsrnr-gsi/fix' fai
led
Failed to clone 'fix' a second time, aborting
```

This seems to have begun over the weekend (Saturday) and is still currently an issue.

I have attempted to update my SSH keys and that also has not resolved the problem.

**Comment by MichaelLueken-NOAA** [comment]
Hi @HenryWinterbottom-NOAA,

Beginning at 5 PM on Friday, May 21, VLab transitioned to the Cloud.  Please change your .ssh/config file to point to:

vlab.noaa.gov

rather than

vlab.ncep.noaa.gov

Please let me know if this corrects the problem.

Thanks,

Mike

**Comment by MichaelLueken-NOAA** [comment]
Hi @HenryWinterbottom-NOAA,

As a follow-up, I have sent Hera a helpdesk ticket.  According to known issues, the firewall on Hera is apparently blocking the new site.  I have requested that Hera sys admins check the firewall to allow VLab's new site:

vlab.noaa.gov

to be accessible on Hera.

Mike

**Comment by RussTreadon-NOAA** [comment]
Similar behavior (unable to access vlab.noaa.gov) found on WCOSS_C (Luna/Surge) and Jet.   Tickets submitted to their respective helpdesks requesting vlab.noaa.gov be added to their list of accessible sites.

**Comment by HenryWinterbottom-NOAA** [comment]
@MichaelLueken-NOAA Thank you for following up. I had made the `.ssh/config` update as well without success.

@RussTreadon-NOAA Thank you for the heads up regarding the other RDHPCS machines. 

If either of you learn something new from the helpdesks, please let me know and I will try on RDHPCS Hera and Orion when it returns to service.

**Comment by RussTreadon-NOAA** [comment]
**Jet update**
Access to Vlab gerrit from Jet has been restored.  I was able to successfully execute "git clone --recursive git@github.com:NOAA-EMC/GSI.git" from my user account on Jet.

**WCOSS_C update**
Discussions between VLab, WCOSS, and Cray admins suggest that the connectivity problem is due to issues with various SSH versions.  I do not have an eta as to when this issue will be resolved.  The WCOSS ticket number is 1020793.

**Comment by HenryWinterbottom-NOAA** [comment]
@MichaelLueken-NOAA @RussTreadon-NOAA Thank you for following up.

The issue was resolved on RDHPCS Hera as of last evening (0100 UTC).

I will close this issue.

**Comment by RussTreadon-NOAA** [comment]
Thank you, Henry, for reporting this problem.   As we still have gerrit connectivity issues on WCOSS_C, I would like to reopen this issue.

**Comment by HenryWinterbottom-NOAA** [comment]
@RussTreadon-NOAA Sorry about that -- I missed the WCOSS part in my haste.

**Comment by RussTreadon-NOAA** [comment]
Not a problem.   I'm glad Hera is now working for your an others.   Let's hope Luna/Surge connectivity is also restored in the near future.

**Comment by RussTreadon-NOAA** [comment]
Received the following update from NCO admins via Vlab admins:

_Hello,
Yes, we do have issues and it appears to be network related. Trying to work with one of our Linux admins to get some tools installed on a VM for some testing scenarios. One of the things I find odd is that when testing, vlab.noaa.gov seems to be load balanced between 205.156.8.150 and 205.156.8.170. The first one (150) seems open whereas 170 always seems closed for tcp/29418. Is traffic to 170 handled differently? Both are open for 443. Oddly enough, even though 29418 is open I still can't "git clone" using tcp/29418. I even fixed my host file to point directly to  205.156.8.150 (the working one). Still doing some additional tests but some odd inconsistencies._

6/1/2021 execution of `git clone --recursive https://github.com/NOAA-EMC/GSI.git .` on Luna does not successfully complete.   Attempts to connect to gerrit to extract `fix/` and `libsrc/` fail.

**Comment by ShelleyMelchior-NOAA** [comment]
Thanks for the update, Russ.

**Comment by aerorahul** [comment]
@MichaelLueken-NOAA @arunchawla-NOAA 
Is there a plan and timeline to move `GSI-fix` and `GSI-libsrc` to a public domain that will not need authentication?

**Comment by MichaelLueken-NOAA** [comment]
@aerorahul 
No, there are no plans to move away from VLab for GSI-fix at this time.

Until the latest version of the wrfio library is available on all machines, GSI-libsrc will be required.  So, once operational systems are no longer run on the Crays, the use of GSI-libsrc will no longer be required.

**Comment by arunchawla-NOAA** [comment]
Why dont we follow the process for fix files that we do for ufs-weather-model at this time. (i.e. stage it for now in the different platforms). We are looking for a better solution with AWS but can we not do this temp solution till we have a more permanent solution ?

**Comment by MichaelLueken-NOAA** [comment]
@arunchawla-NOAA 
Arun, this isn't a real solution.  The fix files are maintained by GSI developers.  So, all GSI developers will need to be able to access these files and change them.  Changes to the GSI fix files need to be versioned (outside of binary fix files).  External developers will need the fix files and won't be able to access these files unless they are located in a repository somewhere.  So, there is no reason to change the method that the GSI maintains the fix files at this time.

**Comment by arunchawla-NOAA** [comment]
So these are non binary fix files ? Presumably then even an AWS storage will not help. Are your binary fix files also in the repo or do you store those separately ? How often are these fix files changed?

**Comment by MichaelLueken-NOAA** [comment]
@arunchawla-NOAA 
All fix files are in VLab (binary and non-binary).  The convinfo and satinfo files are changed anytime new observations are added, when data goes bad, and when data is once again deemed correct.  So, the non-binary files can change from weekly to monthly.  All fix files need to be maintained together to ensure that external developers have access to them (it's not difficult to get users an account on VLab).

**Comment by aerorahul** [comment]
@MichaelLueken-NOAA 
Can you tell us which tag of NCEPLIBS-wrf_io from [here](https://github.com/NOAA-EMC/NCEPLIBS-wrf_io) has been tested with the GSI?
We can get that tag installed and made available as a module on all machines (including Cray).

**Comment by MichaelLueken-NOAA** [comment]
@aerorahul
The tag of NCEPLIBS-wrf_io that works with the GSI is the latest, v1.2.0, from May 21.

**Comment by aerorahul** [comment]
@MichaelLueken-NOAA 
A request has been made to install NCEPLIBS-wrf_io v1.2.0 via the stack [here](https://github.com/NOAA-EMC/hpc-stack/issues/288)

**Comment by aerorahul** [comment]
@MichaelLueken-NOAA 
NCEPLIBs-wrf_io v1.2.0 is installed via the stack.
Please let us know when the GSI-libsrc can be removed from this repository submodule list.

We are now working on finding a solution for the GSI-fix files.

**Comment by arunchawla-NOAA** [comment]
@MichaelLueken-NOAA can you confirm if NCEPLIBS-erf_iov1.2.0 works for you ? 

We need to minimize the complexities in our repos as much as possible. I do not think access to VLAB through the crays will happen anytime soon

**Comment by MichaelLueken-NOAA** [comment]
@aerorahul @arunchawla-NOAA I can confirm that the GSI compiles and runs with bit reproducibility using NCEPLIBS-wrf_io v1.2.0.

What is NCO's rule with respect to loading modules for production builds?  I don't see wrfio-intel/1.2.0 on the NCO production modules on Cray.  Will this need to be addressed?

**Comment by arunchawla-NOAA** [comment]
we will have an hpc-stack v1.2.0 installed by NCO on the new crays soon. Confirm that NCEPLIBS-wrf_iov1.2.0 is available, if not we can open a ticket

**Comment by aerorahul** [comment]
@MichaelLueken-NOAA It is available on WCOSS2

https://github.com/NOAA-EMC/hpc-stack/wiki/Official-Installations

@Hang-Lei-NOAA Can confirm the path if there has been an update since the release of hpc-stack v1.2.0.

**Comment by aerorahul** [comment]
There is also an updated installation of hpc-stack on WCOSS2 since the release v1.2.0 at: 
```
/lfs/h1/emc/nceplibs/noscrub/hpc-stack/libs/hpc-stack/modulefiles/stack
```

**Comment by MichaelLueken-NOAA** [comment]
I need the official NCO production module of wrf_io/1.2.0 on WCOSS2.  Once it is available, I will remove GSI-libsrc.

**Comment by aerorahul** [comment]
There is no such thing on WCOSS2. 
There is no production or experiments on WCOSS2.
There is no general user access on WCOSS2.

**Comment by aerorahul** [comment]
@MichaelLueken-NOAA Do you have an official NCO production module for bufr, nemsio, etc on WCOSS2?

**Comment by MichaelLueken-NOAA** [comment]
@aerorahul I don't see what is so difficult to understand.  When I use:

`module avail`

If the wrf_io/1.2.0 module isn't available, I won't remove GSI-libsrc.  The machines are new, now is the time to get an official version of all libraries on them.  I will not add:

`/lfs/h1/emc/nceplibs/noscrub/hpc-stack/libs/hpc-stack/modulefiles/stack`

to the modulefile for WCOSS2.  There should be no reason to.

Again, once wrf_io/1.2.0 is available and can be loaded directly on WCOSS2, without needing to add a separate modulefile path, I will remove GSI-libsrc.

**Comment by aerorahul** [comment]
@Hang-Lei-NOAA 
@MichaelLueken-NOAA wants these libraries to be installed by NCO in NCO maintained paths.
They won't accept the installations from the NCEPLibs team even if the GSI is bit-reproducible.
I don't know why that is the case for WCOSS2, when they are perfectly fine with this approach on WCOSS_D
https://github.com/NOAA-EMC/GSI/blob/99346e4c5dbb6487b7595e8b600b1e0ed5cd4194/modulefiles/modulefile.ProdGSI.wcoss_d#L21

@MichaelLueken-NOAA Please open tickets with NCO, since it is not something we can help.

Until then the GSI will retain its submodules and this work will take a back seat.

**Comment by MichaelLueken-NOAA** [comment]
@aerorahul Yeah, the GFS v16 implementations don't use that modulefile, they use:

https://github.com/NOAA-EMC/GSI/blob/3616ecb1bd72000f3bbf502012a1b115fabb7ec5/modulefiles/modulefile.ProdGSI.wcoss_d

The upcoming implementation will use:

https://github.com/CatherineThomas-NOAA/GSI/blob/3616ecb1bd72000f3bbf502012a1b115fabb7ec5/modulefiles/modulefile.ProdGSI.wcoss_d

No module uses.  Just wanted to clear this up.

**Comment by aerorahul** [comment]
It is not in the `master`, and there is no open PR or an issue with this upcoming change.
There is no way to see what is coming down the pipe, and no one can track down individual forks and the branches within those forks.

This is beyond us, since we are not going to be able to install the library in the "official" path.
Please take any necessary action to that end, and let us know an estimated date of when to expect the GSI-libsrc is removed from the submodule list.

**Comment by TingLei-daprediction** [comment]
Hi, @RussTreadon-NOAA  May I know what is the status for connecting to vlab on wcoss_c?
This morning (01/11/2022), I still get the error message" ssh: connect to host vlab.noaa.gov port 29418: Connection timed out" on surge.  Thanks

**Comment by RussTreadon-NOAA** [comment]
6/22/2021 email from NCO with following guidance

`recommend EMC continues to use Mars/Venus and then move the code to Luna/Surge in the meantime.`

WCOSS_C will be decommissioned summer 2022.   Closing this issue.

**Comment by TingLei-daprediction** [comment]
@RussTreadon-NOAA  Thanks for the udpate

**Comment by aerorahul** [comment]
> I need the official NCO production module of wrf_io/1.2.0 on WCOSS2. Once it is available, I will remove GSI-libsrc.

@MichaelLueken-NOAA 
NCO has installed `wrf_io/1.2.0` on WCOSS2.
It can be accessed as:
```
module load PrgEnv-intel/8.1.0 intel/19.1.3.304
module load craype/2.7.10 cray-mpich/8.1.9
module load wrf_io/1.2.0
```
Does this satisfy the requirements to remove GSI-libsrc submodule?

**Comment by MichaelLueken-NOAA** [comment]
@aerorahul Yes, since wrf_io/1.2.0 is available on WCOSS2, I can begin the process of removing GSI-libsrc (ultimately, it is up to the DA review committee to remove the submodule, but I don't see any reason that they would want to keep it).  I will send an email to the review committee now to see if they have issues with it now.
